stackslib/src/net/api/blocksimulate.rs

```diff
@@ -0,0 +1,380 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use clarity::vm::costs::ExecutionCost;
+use clarity::vm::Value;
+use regex::{Captures, Regex};
+use stacks_common::codec::StacksMessageCodec;
+use stacks_common::types::chainstate::{BlockHeaderHash, ConsensusHash, StacksBlockId, TrieHash};
+use stacks_common::types::net::PeerHost;
+use stacks_common::util::hash::Sha512Trunc256Sum;
+use stacks_common::util::secp256k1::MessageSignature;
+
+use crate::burnchains::Txid;
+use crate::chainstate::nakamoto::miner::NakamotoBlockBuilder;
+use crate::chainstate::nakamoto::{NakamotoBlock, NakamotoChainState};
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::events::TransactionOrigin;
+use crate::chainstate::stacks::miner::{BlockBuilder, BlockLimitFunction, TransactionResult};
+use crate::chainstate::stacks::{Error as ChainError, StacksTransaction, TransactionPayload};
+use crate::net::http::{
+    parse_json, Error, HttpNotFound, HttpRequest, HttpRequestContents, HttpRequestPreamble,
+    HttpResponse, HttpResponseContents, HttpResponsePayload, HttpResponsePreamble, HttpServerError,
+};
+use crate::net::httpcore::{RPCRequestHandler, StacksHttpResponse};
+use crate::net::{Error as NetError, StacksHttpRequest, StacksNodeState};
+
+#[derive(Clone)]
+pub struct RPCNakamotoBlockSimulateRequestHandler {
+    pub block_id: Option<StacksBlockId>,
+    pub auth: Option<String>,
+}
+
+impl RPCNakamotoBlockSimulateRequestHandler {
+    pub fn new(auth: Option<String>) -> Self {
+        Self {
+            block_id: None,
+            auth,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct RPCSimulatedBlockTransaction {
+    pub txid: Txid,
+    pub tx_index: u32,
+    pub data: Option<StacksTransaction>,
+    pub hex: String,
+    pub result: Value,
+    pub stx_burned: u128,
+    pub execution_cost: ExecutionCost,
+    pub events: Vec<serde_json::Value>,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct RPCSimulatedBlock {
+    pub block_id: StacksBlockId,
+    pub block_hash: BlockHeaderHash,
+    pub block_height: u64,
+    pub parent_block_id: StacksBlockId,
+    pub consensus_hash: ConsensusHash,
+    pub fees: u128,
+    pub tx_merkle_root: Sha512Trunc256Sum,
+    pub state_index_root: TrieHash,
+    pub timestamp: u64,
+    pub miner_signature: MessageSignature,
+    pub signer_signature: Vec<MessageSignature>,
+    pub transactions: Vec<RPCSimulatedBlockTransaction>,
+    pub valid_merkle_root: bool,
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCNakamotoBlockSimulateRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v3/blocks/simulate/(?P<block_id>[0-9a-f]{64})$"#).unwrap()
```

Comment:
Do you think the word "replay" better captures the intent of this API than "simulate"?  The code only lets you re-execute an already-processed block; it doesn't let you execute an arbitrary block's transactions.